### ISSUE TITLE
buf_deref: links < 0 (-1) in Daynaport driver

### ISSUE THREAD

**Issue by czietz** [freemint/freemint#342]
Reported here:
https://forum.atari-home.de/index.php/topic,17912.msg269536.html#msg269536
User gets the error message mentioned above when using FreeMiNT's Daynaport driver (with BlueSCSI's Daynaport emulation).

This error message happens when a network packet buffer is freed twice:
https://github.com/freemint/freemint/blob/d01c39f4db65907dcafe8428d294669ad75d0212/sys/sockets/buf.c#L451-L454

I had a brief look at the Daynaport driver and found something that looks sketchy to me. But I don't have time for further investigation, and no hardware to test with. 

**Can someone please take over?**

In the Daynaport driver, when if_enqueue() fails, the buffer is freed by calling buf_deref():
https://github.com/freemint/freemint/blob/d01c39f4db65907dcafe8428d294669ad75d0212/sys/sockets/xif/daynaport/scsilink.c#L420-L423 
However, the buffer is _already_ freed in if_enqueue(), making this a double free:
https://github.com/freemint/freemint/blob/d01c39f4db65907dcafe8428d294669ad75d0212/sys/sockets/inet4/if.c#L52-L65

**Comment by th-otto** [comment]
> **Can someone please take over?**

I can't test it, but your investigation is certainly correct, the buffer is freed twice. I think we can just remove the call to buf_deref in scsilink.c, that should fix atleast the panic.

The other question is *why* that happens. It is certainly suspoicious that the buffer queue overruns.

**Comment by th-otto** [comment]
Pushed a fix for it. Same things happened also at 3 other places.

**Comment by czietz** [comment]
> Pushed a fix for it. Same things happened also at 3 other places.

Thank you!

So, @franklukas, you could test if SCSILINK.XIF from the latest snapshot with Thorsten's fix (https://tho-otto.de/snapshots/freemint/bootable/) get you further along.

**Comment by mikrosk** [comment]
> The other question is why that happens. It is certainly suspicious that the buffer queue overruns.

Indeed. I know for a fact that there is a couple of users of Daynaport / PiSCSI / BlueSCSI who never experienced (or at least report) such a  crash.

**Comment by th-otto** [comment]
There seems to be some other things that needs checking of buffer (de-)references. For example

- fec_output https://github.com/freemint/freemint/blob/16ece466b308a9ba42a75345387600e3a178c72a/sys/sockets/xif/fec/fec.c#L1270 sometimes calls buf_deref on the original buffer when it returns an error, sometimes not.
- in several cases, bpf_input() was already called, but then the functions may return an error. IMHO bpf processing should not be done on buffers that are discarded later, but there is no easy way to remove it again from the bpf queue, since it might have been merged with another buffer
- generally, that bpf handling looks a bit broken to me. It's purpose is to do filtering of the packets, but in a lot of cases, the original buffer is send directly unfiltered to the hardware nevertheless.

**Comment by franklukas** [comment]
> > Pushed a fix for it. Same things happened also at 3 other places.
> 
> Thank you!
> 
> So, @franklukas, you could test if SCSILINK.XIF from the latest snapshot with Thorsten's fix (https://tho-otto.de/snapshots/freemint/bootable/) gets you further along.

I run the XIF (8679 Bytes) and a new error comes up ...

pid   0 (MiNT): buf_free: invalid buf size: 265280 (0)
FATAL ERROR. You must reboot the system.

------------
The original Dayna SCSILINK Hardware works fine. The BlueSCSI has the problem. I use the MiNT 030 package from 7/2024 (Download time) ...

**Comment by th-otto** [comment]
@mikrosk : could you please try to get your BlueSCSI device up and see whats going on? I can only do some wild guesses here.

**Comment by franklukas** [comment]
System: Atari Falcon, DFB1 with 64MB Fastram, patched TOS 4.04 with Fastram support


I tested STING too and it works fine with the original Dayna SCSILINK but not with the BlueSCSI v2 Desktop

**Comment by th-otto** [comment]
Some things that i noticed: in older scsi library sources like https://github.com/ggnkua/Atari_ST_Sources/blob/9549a8f4a5296e3b5c11b0132bc6e56c3b166867/C/Stefan%20Engel/SCSIDRIV/SRC/C/SCSIIO.C#L428 there is a check whether I/O goes do fastram. No such check is made in the library of the daynaport driver in Mint. Maybe such checks are needed atleast on falcon or STE, where the SCSI controller can address only 24bit IIRC.

**Comment by czietz** [comment]
Idk for sure, but I would expect the SCSIDRV driver to handle this in the same way as a hard disk driver will go through the FRB if needed.

**Comment by th-otto** [comment]
Yes, thats possible. EmuTOS, when compiled with CONF_WITH_SCSI_DRIVER, also already handles this. So maybe such a check is unneeded in application code.

**Comment by franklukas** [comment]
The BlueSCSI device with the entries for Wi-Fi looging successfully into my Fritzbox Router. So there is no problem ...

But SCSILINK.XIF and STING are not working. With the original Hardware it works without any problems. My guess is that there are differences between the Macintosh and Atari drivers when dealing with the BlueSCSI Hardware?
